# CRC_MethylationAnalysis
The code is used to analyze the targeted bisulfite sequencing data for colorectal cancer diagnosis.
THe code was consisted of three parts.
1. The codes to combine the CGmap files generated by the BSseeker2. (CGMapMerge.R and CpGFilter.R);
2. CpGsite2Gene_byMean.R: Compute the mean methylation ratio of multiple CpG sites in a gene to represent the methylation level of the gene.
3. MethylationStatus.R: Performs the analysis to show the OR, Confidence Interval, and P.value, Sensitivity, Specificity, AUC of each Gene with logistic regression.
4. MethylationBoxplot.R: plot the boxplot.
5. MethylCurve.R: plot the methylation curve plot.
6. ROCcurve.R: plot the ROC curve. 
7. MLresult.R: Performs a five-fold cross validation of the methylation data using ten different kinds of machine learning methods.
8. ML_classfiers_Table.R: Obtain the results from MLresult.R and generated the mean sensitivity, specificity and accuracy for the training and test data.
9. Subgroup_Compare.R: Compare the diagnostic abilities of the biomarkers in different subgroups.
10. Preliminary_Analysis.Rmd: The main procedures to identify the potential ZFGs for CRC diagnosis.
11. DiscoveryDataAnalysis.Rmd: The main procedures to show the methylation and expression changes of these ZFGs in the discovery dataset.
12. ValidationAnalysis.Rmd: The main procedures for analyzing the validation datasets.
